runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -7176,29 +7176,39 @@ J9::Z::TreeEvaluator::VMifInstanceOfEvaluator(TR::Node * node, TR::CodeGenerator
    }
 
 /**
- * Generates a quick runtime test for valueType node and in case if node is of valueType, generates a branch to helper call
+ * Generates a quick runtime test for valueType node and in case if node is of valueType or valueBased, generates a branch to helper call
  *
  * @param node monent/exit node
+ * @param tempReg temporary register used for loading the constant flag
+ * @param deps register dependency
  * @param mergeLabel Label pointing to merge point
  * @param helperCallLabel Label pointing to helper call dispatch sequence.
  * @param cg Codegenerator object
  * @return Returns a register containing objectClassPointer
  */
 static TR::Register*
-generateCheckForValueTypeMonitorEnterOrExit(TR::Node *node, TR::LabelSymbol* mergeLabel, TR::LabelSymbol *helperCallLabel, TR::CodeGenerator *cg)
+generateCheckForValueMonitorEnterOrExit(TR::Node *node, TR::Register *tempReg, TR::RegisterDependencyConditions *deps, TR::LabelSymbol* mergeLabel,
+                                        TR::LabelSymbol *helperCallLabel, TR::CodeGenerator *cg)
    {
    TR::Register *objReg = cg->evaluate(node->getFirstChild());
    TR::Register *objectClassReg = cg->allocateRegister();
 
    TR::TreeEvaluator::genLoadForObjectHeadersMasked(cg, node, objectClassReg, generateS390MemoryReference(objReg, TR::Compiler->om.offsetOfObjectVftField(), cg), NULL);
 
-   // Currently the value of J9ClassIsValueType flag maps to the second (low order) byte of the class flags field. To
-   // be able to use the TM instruction we need to ensure we safe guard against changes of this value since we have to
-   // hardcode the offset from the class flags field to reach the byte corresponding to the bit we want to test.
-   static_assert(static_cast<uint32_t>(J9ClassIsValueType & 0xFF00) == static_cast<uint32_t>(J9ClassIsValueType), "Expecting J9ClassIsValueType to be within second byte of classFlags");
+   TR::MemoryReference *classFlagsMemRef = generateS390MemoryReference(objectClassReg, static_cast<uint32_t>(static_cast<TR_J9VMBase *>(cg->comp()->fe())->getOffsetOfClassFlags()), cg);
 
-   TR::MemoryReference *classFlagsMemRef = generateS390MemoryReference(objectClassReg, static_cast<uint32_t>(static_cast<TR_J9VMBase *>(cg->comp()->fe())->getOffsetOfClassFlags()) + 2, cg);
-   generateSIInstruction(cg, TR::InstOpCode::TM, node, classFlagsMemRef, J9ClassIsValueType >> 8);
+   // J9_CLASS_DISALLOWS_LOCKING_FLAGS is currently 0x40400 (J9ClassIsValueType | J9ClassIsValueBased)
+   if(cg->comp()->target().is64Bit())
+      {
+      // TODO: should deps be passed to genLoadLongConstant 
+      genLoadLongConstant(cg, node, J9_CLASS_DISALLOWS_LOCKING_FLAGS, tempReg, NULL, deps, NULL);
+      generateRXInstruction(cg, TR::InstOpCode::NG, node, tempReg, classFlagsMemRef);
+      }
```

Comment:
Yes you should pass the dependency. Please add a space following `if` and I'd remove the comment regarding `J9_CLASS_DISALLOWS_LOCKING_FLAGS` as it poses one more place that needs to get updated should that value change. Users can easily look up the value if they need to:
https://github.com/eclipse/openj9/blob/b981493d215c29b078f3044dfd8b6370207771ae/runtime/oti/j9.h#L312

Reply:
Couple of changes
1. When you are anding the classFlags field with the constant value, I think you are using wrong set of instructions for 64 Bit. It should be same as 32 bit as classFlags within J9Class is 32-Bit field [1]. So the current comparison would be incorrect as it will AND the `J9_CLASS_DISALLOWS_LOCKING_FLAGS` with next field in J9Class after `classFlags`. I would just remove the check for platform size here. 
2. Also I do not think so you need to pass in tempReg as only use of it is before branch and is not used within internal control flow afterwards, so IMO you can just allocate one in this routine and stop using it before returning. 
3. I do not think you need to pass `deps` to any of the routine as you are loading a constant value which is not an address and we would not have to use literal pool to store the immediate field into register and agains the load is not within internal control flow.


[1]. https://github.com/eclipse/openj9/blob/b981493d215c29b078f3044dfd8b6370207771ae/runtime/oti/j9nonbuilder.h#L3080